.changeset/thin-oranges-boil.md

```diff
@@ -0,0 +1,15 @@
+---
+"@remix-run/react": patch
+---
+
+Fixed a problem with live reload and firefox infinitely reloading the page
+
+The problem is:
+
+1. Firefox is calling `ws.onclose` immediately upon connecting (?!)
+2. Then we’re trying to reconnect, and upon reconnection, we reload the page.
+3. Firefox then calls `ws.onclose` again after reconnecting and the loop starts over
+
+This fix is to check `event.code === 1006` before actually trying to reconnect and the reload the page. 1006 means the connection was closed abnormally (https://www.rfc-editor.org/rfc/rfc6455#section-7.4.1). In our case, that means the server was shut down in local dev and then the socket can reconnect again when the server is back up.
+
+It’s unclear to me why Firefox is calling `onclose` immediately upon connecting to the web socket, but it does.
```

Comment:
Hi. When I first had this issue I did some digging and might have an idea of what’s happening:

1. files change
2. [Chrome, Firefox] websocket receives message `type: RELOAD`, triggers `window.location.reload`
3. [Chrome, Firefox] start reload _Navigated to http://localhost:3000/_
4. [Firefox] closes the websocket connection `code: 1001`
5. [Firefox] sets timeout to call `remixLiveReloadConnect`
6. [Firefox] timeout completes _before_ reload is complete and calls:
  `remixLiveReloadConnect({ onOpen() { /* … */ } })`
7. [Firefox] re-connects to the websocket, calls `onOpen` method that triggers another `window.location.reload` starting the infinite reloading

My conclusion is Firefox does **not** in fact call `onclose` upon connecting. It is the other way around, the connection is closed and tries to reconnect before the browser fully unloads the page. Both Chrome and Safari seem to only send the websocket close event when the dev server is stopped and not when the page reloads.

The reason why the `onOpen` config was introduced was to enable automatic [re-connection when the dev server is stopped and restarted](https://github.com/remix-run/remix/pull/859). When that happens the CloseEvent is sent with code `1006` which is why [both of us](https://github.com/remix-run/remix/issues/2997#issuecomment-1277308621) came up with the `event.code === 1006` check.

This fixes the infinite reloads. But the issue on Firefox is actually a race condition. When the timeout is increased (e.g. to `10000`) the page reload reliably finishes before the timeout re-initializes, re-connects to the websocket and the reloads loop kicks in.